Control shared/static linking via BUILD_SHARED_LIBS and ament_cmake_ros

esteve opened the pull request:
This PR delegates the decision to build a shared or a static library to CMake's BUILD_SHARED_LIBS

Update:

Connects to https://github.com/ros2/ros2/issues/306

dirk-thomas commented on rmw_fastrtps_cpp/package.xml:
```diff
@@ -7,10 +7,10 @@
   <buildtool_depend>fastrtps_cmake_module</buildtool_depend>
 
-  <buildtool_export_depend>ament_cmake</buildtool_export_depend>
+  <buildtool_export_depend>ament_cmake_ros</buildtool_export_depend>
```
The package doesn't export `ament_cmake_ros` anywhere therefore I don't think it needs to be exported here either.

esteve replied:
Sorry, I forgot about this. I've reverted that change and rebased against master. Thanks!

dirk-thomas commented:
Looks good to me.

Normal CI builds:
* http://ci.ros2.org/job/ci_linux/2344/
* http://ci.ros2.org/job/ci_osx/1800/
* http://ci.ros2.org/job/ci_windows/2446/

CI builds with `-DBUILD_SHARED_LIBS=0`:
* http://ci.ros2.org/job/ci_linux/2345/
* http://ci.ros2.org/job/ci_osx/1801/
* http://ci.ros2.org/job/ci_windows/2447/

dirk-thomas commented:
It looks like the `ament_cmake_ros` package lacks an export directive for `ament_cmake`. Does it compile for you locally?

esteve commented:
IIRC it used to, but haven't tried recently, I only updated this PR and addressed your comment because it was getting stale. Isn't this sufficient to export `ament_cmake` already?

https://github.com/ros2/ament_cmake_ros/blob/master/ament_cmake_ros/package.xml#L12

or you mean in the `CMakeLists.txt` file?

dirk-thomas commented:
Yes, it needs to be exported in the CMake file in order to be part of the generated CMake config file.

esteve commented:
@dirk-thomas I've submitted https://github.com/ros2/ament_cmake_ros/pull/1, but don't review it just yet, I'm building this locally.

esteve commented:
@dirk-thomas thanks for merging https://github.com/ros2/ament_cmake_ros/pull/1 so quickly, but I wasn't done building this yet :-D

dirk-thomas commented:
The change was trivial to review and definitely the right thing to do. The manifest already exported the information. The lack of exporting it in CMake was an obvious oversight.

esteve commented:
@dirk-thomas I know, it was just a tongue-in-cheek comment, buddy.

esteve commented:
@dirk-thomas I've submitted https://github.com/ros/console_bridge/pull/42 so that `console_bridge` can be linked against `class_loader` statically, though I don't know what that usercase would be. IIRC, when I've built ROS 2.0 statically I've skipped `class_loader` altogether, since there's not need for it.

esteve commented:
Regarding the `-fPIC` flag in `console_bridge`, here's the error output from http://ci.ros2.org/job/ci_linux/2345/ (`BUILD_SHARED_LIBS=OFF`)

```
/usr/bin/ld: /home/rosbuild/ci_scripts/ws/install/console_bridge/lib/libconsole_bridge.a(console.cpp.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
16:48:11 /home/rosbuild/ci_scripts/ws/install/console_bridge/lib/libconsole_bridge.a: error adding symbols: Bad value
16:48:11 collect2: error: ld returned 1 exit status
16:48:11 CMakeFiles/class_loader.dir/build.make:174: recipe for target 'libclass_loader.so' failed
16:48:11 make[2]: *** [libclass_loader.so] Error 1
16:48:11 CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/class_loader.dir/all' failed
16:48:11 make[1]: *** [CMakeFiles/class_loader.dir/all] Error 2
16:48:11 Makefile:138: recipe for target 'all' failed
16:48:11 make: *** [all] Error 2
```

dirk-thomas commented:
Normal CI builds look good:
* http://ci.ros2.org/job/ci_linux/2346/
* http://ci.ros2.org/job/ci_osx/1802/
* http://ci.ros2.org/job/ci_windows/2448/

The CI builds with `-DBUILD_SHARED_LIBS=0` use ros/console_bridge#42:
* http://ci.ros2.org/job/ci_linux/2348/
* http://ci.ros2.org/job/ci_osx/1804/
* http://ci.ros2.org/job/ci_windows/2450/

dirk-thomas commented:
The CI build with `-DBUILD_SHARED_LIBS=0` using ros/console_bridge#43:

* http://ci.ros2.org/job/ci_windows/2474/

esteve commented:
@dirk-thomas thanks for triggering the build, @j-rivero 's pull request seems to have fixed the `PIC` issues, fantastic!

dirk-thomas commented:
Waiting for the `console_bridge` PR to be merged.

esteve commented:
@dirk-thomas now that https://github.com/ros/console_bridge/pull/43 is merged, is there anything pending that needs to be addressed? Thanks.

dirk-thomas commented:
This was still blocked by ament/gtest_vendor#4 et al which just got merged. Before running CI on this I am waiting for the comments on the other PRs in order to test them together.

dirk-thomas commented:
This also needs to be rebased since it fails to compile against the current master of other packages.

esteve commented:
@dirk-thomas rebased. IIRC the build farm would rebase before building if there are no conflicts, though.

mikaelarguedas commented:
@esteve Unfortunately the buildfarm rebases only if you give it a branch name to rebase. If you provide it a repos file it just clones the repos and can't rebase because it doesn't know against what branch it should be rebased (discussion [here](https://github.com/ros2/ci/issues/17) if you're interested)

esteve commented:
@mikaelarguedas that's unfortunate indeed, thanks for the explanation!

wjwwood commented:
I ran into the same thing testing https://github.com/ros2/rmw/pull/80, an unfortunate issue with testing forks in our CI.

dirk-thomas commented:
Since this is unrelated to the `rosidl` PRs I triggered some CI builds just for this patch since it looks good as is:

* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2495)](http://ci.ros2.org/job/ci_linux/2495/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=94)](http://ci.ros2.org/job/ci_linux-aarch64/94/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1962)](http://ci.ros2.org/job/ci_osx/1962/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2605)](http://ci.ros2.org/job/ci_windows/2605/)

esteve commented:
@dirk-thomas I just pushed a rebased branch.

dirk-thomas commented:
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2​55​2)](http://ci.ros2.org/job/ci_linux/2​55​2/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=148)](http://ci.ros2.org/job/ci_linux-aarch64/148/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=2​01​9)](http://ci.ros2.org/job/ci_osx/2​01​9/)
* [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2​66​1)](http://ci.ros2.org/job/ci_windows/2​66​1/)

dirk-thomas commented:
CI builds look good. Thanks for iterating on this.

esteve commented:
@dirk-thomas thanks for the patience!
